internal/secp256k1_omni_context/src/keypair.rs

```diff
@@ -0,0 +1,183 @@
+use secp256k1::{
+    self, rand::Rng, Message, PublicKey, RecoverableSignature, Secp256k1, SecretKey, Signature,
+};
+use std::str::FromStr;
+
+#[derive(Debug)]
+pub enum Error {
+    NoSecretKeyProvided,
+    MultipleSecretKeysProvided,
+    Secp256k1(secp256k1::Error),
+}
+
+impl From<secp256k1::Error> for Error {
+    fn from(err: secp256k1::Error) -> Self {
+        Error::Secp256k1(err)
+    }
+}
+
+#[derive(Debug)]
+pub struct Builder {
+    secp: Secp256k1<secp256k1::All>,
+    secret_key_slice: Option<Vec<u8>>,
+    secret_key_hex: Option<String>,
+    secret_key: Option<SecretKey>,
+}
+
+impl Builder {
+    pub fn new(secp: Secp256k1<secp256k1::All>) -> Builder {
+        Builder {
+            secp,
+            secret_key_slice: None,
+            secret_key_hex: None,
+            secret_key: None,
+        }
+    }
+
+    pub fn secret_key(self, secret_key: SecretKey) -> Builder {
+        Builder {
+            secret_key: Some(secret_key),
+            ..self
+        }
+    }
+
+    pub fn secret_key_slice(self, data: &[u8]) -> Builder {
+        let vec = data.to_vec();
+        Builder {
+            secret_key_slice: Some(vec),
+            ..self
+        }
+    }
+
+    pub fn secret_key_hex(self, key: &str) -> Builder {
+        Builder {
+            secret_key_hex: Some(key.to_string()),
+            ..self
+        }
+    }
+
+    pub fn rng<R: Rng>(self, rng: &mut R) -> Builder {
+        Builder {
+            secret_key: Some(SecretKey::new(rng)),
+            ..self
+        }
+    }
+
+    pub fn build(self) -> Result<KeyPair, Error> {
+        let mut secret_keys_count = 0;
```

Comment:
:wrench: I think you could represent all the possibilities with an enum or even just transforming the slice or hex into a `SecretKey` when the `secret_key_slice` or `secret_key_hex` method is called.

Reply:
> even just transforming the slice or hex into a SecretKey when the secret_key_slice or secret_key_hex method is called.

I'd suggest doing that and simply returning a `Result<Builder, Error>` from each of these. It is still chainable if you `?` inbetween.

Reply:
hum, I see what you mean and I wondered the same.
However, I designed it this way because when using a builder, I expect to be able to chain initialisation parameters and then call `Builder::build()` and only expect to worry about a result when building.
Worrying about a result during initialisation of the builder seems weird to me.

Thoughts?

Reply:
How many valid combinations of these build steps are there though?

It seems like they are just different constructors, where all of them should yield a `Result<KeyPair, Error>`.

Reply:
ok, I'll implement this comment in the next PR.